Fix CurlHandler handling of redirects

stephentoub opened the pull request:
- Relative urls in Location headers were not being handled correctly.  We now allow any url that TryCreate will parse using the current url as a base.
- We were only handling a subset of redirect status codes as redirects, trying to second-guess which status codes libcurl would redirect for, e.g. 300 wasn't being treated as a redirect with regards to our processing of it (it would redirect but we wouldn't reconfigure some of our state). We now let libcurl do all the work.
- We should have been playing it safer with regards to clearing out credentials if we weren't sure a Uri redirect location was valid. Now we do.
- We were tracking state unnecessarily that we should have instead been
  asking libcurl for, e.g. the last url used in a redirect.  This also caused us to store the wrong RequestMessage.Uri in some cases where a location header came in but we didn't actually follow it.

cc: @davidsh, @ericeil, @JunTaoLuo, @bartonjs 
Fixes https://github.com/dotnet/corefx/issues/8941

JunTaoLuo commented:
I have verified that this fixes the repro given in the issue and the MusicStore tests that we were trying to run. cc @Tratcher @muratg

stephentoub commented:
> I have verified that this fixes the repro given in the issue and the MusicStore tests that we were trying to run.

Great, thanks for verifying.

muratg commented:
cc @Tratcher

bartonjs commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs:
```diff
@@ -722,10 +722,22 @@ internal void SetTaskOffsetCount(Task<int> task, int offset, int count)
+                        _requestMessage.RequestUri = finalUri;
+                        return;
+                    }
+                }
```
Do you want a different Debug.Fail for TryCreate failing?

stephentoub replied:
I thought one would be sufficient.  Either we're unable to get a url, or the url we get is bad, and either way we Debug.Fail.

bartonjs replied:
What happens in the "else" case of all this in release? infinite redirect loop? aborted infinite redirect loop? Redirect to the new location but RequestUri stays with the old value?  If the latter, does that run the risk of exposing cookies/creds that shouldn't be?

stephentoub replied:
> What happens in the "else" case of all this in release?

It shouldn't happen, but if it does, the HttpResponseMessage.RequestMessage.RequestUri visible after the request completes will simply be the original Uri requested (or the previous one in the redirect chain if this wasn't the first automatic redirect).

bartonjs replied:
I think I had assumed this was "where are we going", then later I realized it was "where we started"; when I found the actual "where are we going" handler.

bartonjs commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -947,7 +947,7 @@ private static ulong CurlReceiveHeadersCallback(IntPtr buffer, ulong size, ulong
                                     response.Content.Headers.TryAddWithoutValidation(headerName, headerValue);
                                 }
-                                else if (easy._isRedirect && string.Equals(headerName, HttpKnownHeaderNames.Location, StringComparison.OrdinalIgnoreCase))
+                                else if (string.Equals(headerName, HttpKnownHeaderNames.Location, StringComparison.OrdinalIgnoreCase))
```
If a Location header is set on a HTTP-200 response, what happens?  curl calls it an error? we redirect?

What about the Windows version?

(It'd make an interesting test case :smile:)

bartonjs replied:
I was looking to see what the HTTP 1.1 spec had to say about this sort of case.  In most things which talk about Location it is a "go here", but a HTTP 305 uses Location for "use this proxy".

https://tools.ietf.org/html/rfc2616#section-10.3.6.

So we should either still opt-in the redirecting codes, opt-out the 305, or decide "proxies are effectively dead, so who cares?"

stephentoub replied:
I've now limited the check to be for 3xx.

stephentoub replied:
I also added a test for non-redirect status codes and a Location header.

bartonjs commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs:
```diff
@@ -736,46 +736,32 @@ private static IOException MapToReadWriteIOException(Exception error, bool isRea
+            {
+                // In case of a bad location, clear out any creds that may have been set, just in case.
+                state.SetCredentialsOptions(default(KeyValuePair<NetworkCredential, CURLAUTH>));
             }
```
We should probably clear cookies in this error path, too.

stephentoub replied:
Done.

bartonjs commented:
Clearing the cookies is important for the same reason clearing the credentials is.  The Location header + status code part I don't have strong opinions on... maybe checking the HTTP 2.0 spec to see what opinions it has could be enlightening.

LGTM with crumbled cookies.

davidsh commented:
>https://tools.ietf.org/html/rfc2616#section-10.3.6.

We do not use this RFC any longer.  It is obsolete.

We only use RFC 723x for HTTP stuff now.
https://tools.ietf.org/html/rfc7230

bartonjs commented:
@davidsh Cool, so that clears the HTTP 305 response.

Looks like 7231 says it's a redirection target for any 3xx status code (and "where I put it" for a 201); so maybe we can check the code is >= 300 and <= 399, and then treat it as a redirect.  Without that check, a PUT or POST might result in a redirect from a 201.

stephentoub commented:
> so maybe we can check the code is >= 300 and <= 399, and then treat it as a redirect. Without that check, a PUT or POST might result in a redirect from a 201.

This PR gets us out of the game of determining whether a particular status code is a redirect.  We leave that up to libcurl now, and the only thing we do with a Location header is essentially say "in case we end up redirecting, reconfigure the easy handle for making a new request".

bartonjs commented:
> the only thing we do with a Location header is essentially say "in case we end up redirecting, reconfigure the easy handle for making a new request".

Gotcha. Fine by me then.

ericeil commented:
This LGTM, though after reading through @bartonjs's comments, I clearly need to study up on HTTP before I feel qualified to sign off on a change like this. 😃

stephentoub commented:
> though after reading through @bartonjs's comments, I clearly need to study up on HTTP before I feel qualified to sign off on a change like this

The most important thing to know is that the current implementation is broken for a variety of redirect scenarios, e.g. today if a server uses a 300 redirect, credentials/cookies/etc. will not be reconfigured for the new request.

stephentoub commented:
Thanks for the reviews, all.

Jeremy, I've addressed your feedback.

And after reading the Location section of 7231, I've also updated the Uri-combining code to properly preserve fragments, which it wasn't previously doing.

bartonjs commented:
LGTM.  Thanks, Steve.

davidsh commented:
>And after reading the Location section of 7231, I've also updated the Uri-combining code to properly preserve fragments, which it wasn't previously doing.

FWIW, preserving fragments is really only needed for web browsers when redirecting to different URL (pages for the web browsers).  In practice, it won't really matter to preserve the fragment because the same data is fetched on the wire.  Fragments are only used in web browsers when displaying HTML pages.

stephentoub commented:
> FWIW, preserving fragments is really only needed for web browsers 

Thanks, David. Do you recommend I remove this part of the handling then?  It looks like the tests I wrote pass on Windows 10 but fail on Windows 8 ("the data is invalid"), which would seem to suggest that WinHttp was updated to support this.  But, you'd know better than I what the right thing to do here is.

davidsh commented:
>It looks like the tests I wrote pass on Windows 10 but fail on Windows 8 ("the data is invalid"), which would seem to suggest that WinHttp was updated to support this. 

Passing the fragment in the URI to native stacks is useless because servers don't care about receiving the fragment portion of path.

It sounds like WinHTTP was updated to better handle stripping off the fragment portion if it is passed into WinHttpOpenRequest/WinHttpSendRequest. I'm not aware of that change but it was probably done as a convenience to API callers to not have to strip off the fragment themselves.

>Do you recommend I remove this part of the handling then?

I think it ok to leave the code as-is.  It's possible that some developers want to see the complete URI in the redirected URI field `HttpRequestMessage.RequestUri`

stephentoub commented:
Test Innerloop OSX Release Build and Test please (build machine went offline)

stephentoub commented:
> It's possible that some developers want to see the complete URI in the redirected URI field  HttpRequestMessage.RequestUri 

I see.  That's actually handled by libcurl.  The only value in doing the extra manipulations (which do have some cost) in the handler itself was in changing the Uri that was passed to things like cookie lookup.  Since the fragment isn't going to affect such things, I've removed those additional fragment-related changes I made.
